packages/adapter-utils/src/utils.ts

```diff
@@ -1184,25 +1184,26 @@ export const getSubtypes = async (types: ObjectType[], validateUniqueness = fals
   const subtypes: Record<string, ObjectType> = {}
 
   const findSubtypes = async (type: ObjectType): Promise<void> => {
-    await awu(Object.values(type.fields)).forEach(async field => {
-      const fieldContainerOrType = await field.getType()
-      const fieldType = isContainerType(fieldContainerOrType)
-        ? await fieldContainerOrType.getInnerType()
-        : fieldContainerOrType
-
-      if (!isObjectType(fieldType) || types.includes(fieldType)) {
-        return
-      }
-      if (fieldType.elemID.getFullName() in subtypes) {
-        if (validateUniqueness && !subtypes[fieldType.elemID.getFullName()].isEqual(fieldType)) {
-          log.warn(`duplicate ElemIDs of subtypes found. The duplicate is ${fieldType.elemID.getFullName()}`)
+    const fieldsTypes = await Promise.all(Object.values(type.fields).map(field => field.getType()))
+    const additionalPropertiesRefType = type.annotations[CORE_ANNOTATIONS.ADDITIONAL_PROPERTIES]?.refType
+    await awu(additionalPropertiesRefType ? [...fieldsTypes, additionalPropertiesRefType] : fieldsTypes).forEach(
```

Comment:
nit: there's a preference for explicit checks
```suggestion
    await awu(additionalPropertiesRefType !== undefined ? [...fieldsTypes, additionalPropertiesRefType] : fieldsTypes).forEach(
```